amethyst_assets/src/source/http.rs

```diff
@@ -0,0 +1,92 @@
+use js_sys::Uint8Array;
+use std::path::{Path, PathBuf};
+use web_sys::{XmlHttpRequest, XmlHttpRequestResponseType};
+
+#[cfg(feature = "profiler")]
+use thread_profiler::profile_scope;
+
+use amethyst_error::{format_err, Error, ResultExt};
+
+use crate::{error, source::Source};
+
+/// HTTP source.
+///
+/// Loads assets inside web worker using XmlHttpRequest.
+/// Used as a default source for WASM target.
+#[derive(Debug)]
+pub struct HTTP {
+    loc: PathBuf,
+}
+
+impl HTTP {
+    /// Creates a new http source.
+    pub fn new<P>(loc: P) -> Self
+    where
+        P: Into<PathBuf>,
+    {
+        HTTP { loc: loc.into() }
+    }
+
+    fn path(&self, s_path: &str) -> PathBuf {
+        let mut path = self.loc.clone();
+        path.extend(Path::new(s_path).iter());
+
+        path
+    }
+}
+
+impl Source for HTTP {
+    fn modified(&self, _path: &str) -> Result<u64, Error> {
+        #[cfg(feature = "profiler")]
+        profile_scope!("http_modified_asset");
+
+        // Unimplemented. Maybe possible to tie into webpack hot module reloading?
+        Ok(0)
+    }
+
+    fn load(&self, path: &str) -> Result<Vec<u8>, Error> {
+        #[cfg(feature = "profiler")]
+        profile_scope!("http_load_asset");
+
+        let path = self.path(path);
+        let path_str = path
+            .to_str()
+            .ok_or_else(|| format_err!("Path contains non-unicode characters: {:?}", path))
+            .with_context(|_| error::Error::Source)?;
+
+        let xhr = XmlHttpRequest::new()
+            .map_err(|_| format_err!("Failed to construct XmlHttpRequest"))
+            .with_context(|_| error::Error::Source)?;
+
+        // Synchronous GET request. Should only be run in web worker.
+        xhr.open_with_async("GET", path_str, false).unwrap();
+        xhr.set_response_type(XmlHttpRequestResponseType::Arraybuffer);
+
+        // We block here and wait for http fetch to complete
+        xhr.send()
+            .map_err(|_| format_err!("XmlHttpRequest send failed"))
+            .with_context(|_| error::Error::Source)?;
+
+        // Status returns a result but according to javascript spec it should never return error.
+        // Returns 0 if request was not completed.
+        let status = xhr.status().unwrap();
+        if status != 200 {
+            let msg = xhr.status_text().unwrap_or_else(|_| "".to_string());
```

Comment:
might save us some debugging time:

```rust
use std::string::ToString; // somewhere on top of the file
let msg = xhr.status_text().unwrap_or_else(ToString::to_string);
```

Reply:
On the second thought, it's the same story as with `status`. According to javascript specification this function should never fail. Even if we want to handle the error, its type is `JsValue` and we don't know what type to cast it to (`ToString` is not implemented for `JsValue` neither is debug). I suggest leaving it as unwrap unless there is a good reason to throw a generic error which is not really helpful.